Feature/78 mock up relay schedule

kevin-wijnen opened the pull request:
Merge into master to allow code inclusion for feature/68.

kevin-wijnen committed:
Old class cleanup

kevin-wijnen committed:
Device Test Class intialized

rlemmers committed:
72 ~ Clean up code and introduce toString

kevin-wijnen committed:
Removal of displayDevice function, clean up toStrings

kevin-wijnen committed:
General cleanup of toStrings

kevin-wijnen committed:
Cleaning up comments, unused code and rebuilt Scheduler toString

kevin-wijnen committed:
JavaDocs added, ActionExecutor cleanup

kevin-wijnen committed:
Change in text layout

kevin-wijnen committed:
Scheduler Test instructions

kevin-wijnen committed:
Task 78: Initial work Scheduler + mock data functionality

Initial work of the desired Scheduler implemented in the proper classes
In unit test for Scheduler > mock data functionality for unit testing

kevin-wijnen committed:
Trigger implementation for future switching moment calculations

kevin-wijnen committed:
Safeguard for switching moments calculation mishaps at boot

kevin-wijnen committed:
Transfer mock up code to different branch

kevin-wijnen committed:
Transfer mock up feature from different branch

kevin-wijnen committed:
Rename function + extra comments

kevin-wijnen committed:
Switching Moment Calculator tests rework

kevin-wijnen committed:
Switching Moment Calculator rework

Implementation of calculating switching moments for two days
Implementation of multiple week days (instead of just a date)
Daily (at midnight) switching moment calculation implemented
General reorganization

kevin-wijnen committed:
Added 'Every Day'  unit test + fixes for missing relays and schedules

kevin-wijnen committed:
Slight correction in comments

rlemmers commented on iec61850serversimulator/pom.xml:
```diff
@@ -55,6 +55,22 @@
+
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-web</artifactId>
```
Is this dependency needed? Normally it's needed when you have a GUI or REST interface.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/ActionExecutor.java:
```diff
@@ -3,57 +3,54 @@
-	//private static final String PRINT_SERVER_MODEL_KEY_DESCRIPTION = "print server's model";
 
-    public ActionExecutor(ServerSap serverSap, ServerModel serverModel, Device device) {
+	public ActionExecutor(final ServerSap serverSap, final ServerModel serverModel, final Device device) {
```
The line starts with a tab. Replace it by 4 spaces and update the settings of your IDE.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/ActionExecutor.java:
```diff
@@ -3,57 +3,54 @@
+				break;
+
+			case DEVICE_SHOW_MODEL:
+				logger.info(this.device.toString());
```
Always include a "break", even when it's the last case.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/ActionExecutor.java:
```diff
@@ -3,57 +3,54 @@
-	
-	
+
 }
```
Always end with a newline.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Clock.java:
```diff
@@ -14,161 +14,208 @@
-			String dataAttribute = bda.getName();
-			
-			switch(dataAttribute) {
+	// If multiple IP addresses are in use, use ; to separate them
```
Either add quotes ("," and ";") or write "... use a comma, otherwise use a semicolon to separate ...."

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Device.java:
```diff
@@ -47,12 +38,33 @@ public void displayDevice() {
-			this.relays[relayNr].displayRelay();
-			for (int scheduleNr = 0; scheduleNr < 50; scheduleNr++) {
-				logger.info(this.getRelay(relayNr + 1).getSchedule(scheduleNr + 1).toString());
+			this.relays[relayNr].toString();
```
This doesn't do anything. If you want to print out the information: logger.info(this.relays[relayNr].displayRelay())

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Device.java:
```diff
@@ -47,12 +38,33 @@ public void displayDevice() {
+	public String toString() {
+		final StringBuilder deviceStringBuilder = new StringBuilder();
 
+		deviceStringBuilder.append("** Printing device:\n\n");
```
Using System.lineSeparator() is better than "\n", because it's replaced by the default of the operating system.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/EventDataListener.java:
```diff
@@ -310,11 +335,14 @@ public EventDataListener(final Device device) {
+			logger.info("Schedules are modified! Calculate switching moments...");
+			this.scheduler.switchingMomentCalculation(this.device);
+			// Initializing scheduler?
+			// scheduler.switchingMomentCalculation(this.device);
```
Can this be deleted?

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/EventDataListener.java:
```diff
@@ -24,22 +27,35 @@
 		// OnzeScheduleScheduler scheduler = new ...;
 	}
 
+	@Scheduled(cron = "0 0 0 * * ?")
```
We can leave this here for now and look at it later. Starting a job according to a cron expression isn't part of the "single responsibility" for the EventDataListener class.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Relay.java:
```diff
@@ -47,66 +43,37 @@ public Relay(ModelNode relayInfo, ModelNode scheduleInfo) {
-				+ this.scheduleInfo + ", schedules=" + Arrays.toString(this.schedules) + "]";
+		final StringBuilder relayStringBuilder = new StringBuilder();
+		relayStringBuilder.append("Relay " + this.indexNumber + "'s light status is " + this.lightStatus + "\n");
+		return relayStringBuilder.toString();
```
Use plain String concatenation here to make it more readable. There's just a small number of parameters, which means the StringBuilder doesn't offer a significant performance difference.

(General remark: when using a StringBuilder, don't use the "+" operator. That's an operator for Strings, so you should replace it by "append" calls.)

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/ServerSimulator.java:
```diff
@@ -25,94 +29,93 @@
+		try {
+			cliParser.parseArguments(args);
+		} catch (final CliParseException e1) {
+			System.err.println("Error parsing command line parameters: " + e1.getMessage());
```
Don't use any System.err.println / System.out.println statements, when you have a logger available.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMoment.java:
```diff
@@ -0,0 +1,71 @@
+ * The destination relay's number
+ * The date and time when to trigger the switching moment (LocalDateTime)
+ * The type of action what to trigger (On = true, Off = false)
+ * @formatter:on
```
In Javadoc, use html-tags to format the data. Here you could make a list using ul/ol/li tags. Only use the @formatter in special cases, when there's really no other option.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculator.java:
```diff
@@ -1,24 +1,202 @@
-		this.device = device;
-	}
+		for (int relayNr = 1; relayNr <= 4; relayNr++) {
+			try {
```
1. Make the error handling simpler by not nesting the try statements.
2. When an exception occurs, the calculateSwitchingMoments code should not continue. Make sure you catch the exception and then throw your own exception. The caller of calculateSwitchingMoments should catch and log those exceptions,

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculator.java:
```diff
@@ -1,24 +1,202 @@
+		return switchingMoments;
 	}
 
+	public SwitchingMoment calculateSwitchingMoment(int relayNr, Schedule schedule, boolean triggerAction,
```
Should this method be public or private?

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculator.java:
```diff
@@ -1,24 +1,202 @@
+		switch (day) {
+		// Check if day is within 2 days. Always recalculate every day.
+		// TODO: Implementing every day, every week day and every weekend day
+		case 0:
```
Don't switch on this integer, convert it to a value of an enumeration. I think you already have a suitable one.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculator.java:
```diff
@@ -1,24 +1,202 @@
+		 * now.plusDays(1); DayOfWeek day1 = now.getDayOfWeek(); DayOfWeek day2 =
+		 * nextday.getDayOfWeek();
+		 */
+		switch (day) {
```
Consider renaming. Currently, it's hard to tell the difference in meaning between "dayOfWeek" and "day". Both are representing a day or set of days in a week.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculator.java:
```diff
@@ -1,24 +1,202 @@
+			}
+			break;
+
+		default:
```
What is the value of "day" in this case?

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculator.java:
```diff
@@ -1,24 +1,202 @@
+			switchingMoment.setTriggerTime(triggerTime);
+		}
+
+		else if (!triggerAction) {
```
In the else, you already know triggerAction is false. So simplify the structure to:
if (triggerAction) {
...
} else { 
...
}

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculator.java:
```diff
@@ -1,24 +1,202 @@
+			DateTimeFormatter dateFormat = DateTimeFormatter.ofPattern("yyyyMMdd");
+			String stringDate = String.valueOf(day);
+			LocalDateTime triggerDate = LocalDateTime.parse(stringDate, dateFormat);
+			switchingMoment.setTriggerTime(triggerDate);
```
We'll revisit this code later. Formatting/converting should be done when receiving a value.

rlemmers commented on iec61850serversimulator/pom.xml:
```diff
@@ -55,6 +55,22 @@
+			<groupId>org.postgresql</groupId>
+			<artifactId>postgresql</artifactId>
+			<scope>runtime</scope>
+		</dependency>
```
This PR is about mocks. It should not be mixed with other work.

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -0,0 +1,224 @@
+		return schedule;
+	}
+
+	public Schedule getMockSchedule(int scheduleNr, int relayNr, int dayInt, int fixedTimeInt, int fixedTimeOn,
```
Too many parameters!

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -0,0 +1,224 @@
+			Relay[] relays = new Relay[1];
+			relays[0] = relay;
+			this.device.setRelays(relays);
+		}
```
The else-part is never used.

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -0,0 +1,224 @@
+		SwitchingMoment switchingMomentNextDayOn = switchingMoments.get(1);
+		SwitchingMoment switchingMomentNextDayOff = switchingMoments.get(3);
+
+		LocalDateTime lunchTime = LocalDateTime.now().withHour(12).withMinute(00).withSecond(0).withNano(0);
```
To make this shorter: LocalDateTime.now().truncatedTo(...).withHour(12).

rlemmers replied:
To make it more readable: add "expected" as prefix to the LocalDateTime's you expect.

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -0,0 +1,224 @@
+
+		// Tests
+
+		SwitchingMoment switchingMomentOn = switchingMoments.get(0);
+		SwitchingMoment switchingMomentOff = switchingMoments.get(2);
+		SwitchingMoment switchingMomentNextDayOn = switchingMoments.get(1);
+		SwitchingMoment switchingMomentNextDayOff = switchingMoments.get(3);
```
To make it more readable: add "actual" as prefix.

kevin-wijnen committed:
Configuration removed, fixed crash with it

kevin-wijnen committed:
Pull Request feedback rework

- Unnecessary dependencies and commented code removed
- Tab replaced by 4 spaces (ActionExecutor class)
- Fixed Switch case statement with addition of break (^)
- General text and code formatting improvements
- Added comments to the code that will be revisited later on
- Value of day clarified (although, the code might get revised in the
future to separate the two types of data (re-occurring day or date)

Left to change regarding feedback:
- Migrating Builder to Scheduler class
- Create separate test for custom schedules

kevin-wijnen committed:
Creation of Enums for readability

- Created ScheduleDay enum for clear switch cases (already integrated)
- Created TriggerType enum for clear switch cases (yet to integrate)

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculator.java:
```diff
@@ -70,31 +70,31 @@
-		SwitchingMoment switchingMoment = new SwitchingMoment(relayNr, null, triggerAction);
-		int day = schedule.getDayInt();
+		SwitchingMoment actualSwitchingMoment = new SwitchingMoment(relayNr, null, triggerAction);
+		int scheduledOccurence = schedule.getDayInt();
```
Spelling: scheduledOccurence => scheduledOccurrence

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -17,17 +18,103 @@
+	 * </ul>
 	 */
 
+	public static class MockedScheduleBuilder {
```
This creates a normal Schedule object. Move the "MockedScheduleBuilder" to the Schedule class (and rename it to ScheduleBuilder).

kevin-wijnen committed:
Comment removal/translation to English

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -17,17 +18,103 @@
+			this.scheduleNr = scheduleNr;
+		}
+
+		public MockedScheduleBuilder setRelayNr(int relayNr) {
```
Rename "setXyz" to "withXyz" or "xyz".

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -17,17 +18,103 @@
+			return this;
+		}
+
+		public Schedule buildSchedule() {
```
Create a unit test to prove buildSchedule works correctly.

rlemmers commented on iec61850serversimulator/src/test/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculatorTest.java:
```diff
@@ -62,8 +149,13 @@ public void checkMockSchedule() {
+				.setFixedTimeInt(fixedTimeInt).setFixedTimeOn(fixedTimeOn).setFixedTimeOff(fixedTimeOff)
+				.setTimeOn(timeOn).setTimeOff(timeOff).setBurningMins(burningMinutes).buildSchedule();
+
+		// Schedule schedule = this.getMockSchedule(scheduleNr - 1, relayNr, dayInt,
+		// fixedTimeInt, fixedTimeOn,
+		// fixedTimeOff, timeOn, timeOff, burningMinutes);
```
Remove this commented code.

kevin-wijnen committed:
Switching Moment Calculation error handling implemented

kevin-wijnen committed:
Separated custom schedule test, implemented builder properly

Note: Builder yet to be moved over to Schedule class

kevin-wijnen committed:
Migration of ScheduleBuilder class to Schedule class

kevin-wijnen committed:
Proper naming convention implemented for the Schedule builder

rlemmers commented on iec61850serversimulator/hs_err_pid20174.log:
```diff
@@ -0,0 +1,698 @@
+#
```
Remove this file from the pull request.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Schedule.java:
```diff
@@ -19,38 +19,100 @@
-	public enum DAY {
-		EVERYDAY(0), WEEKDAY(-1), WEEKEND(-2), MONDAY(1), TUESDAY(2), WEDNESDAY(3), THURSDAY(4), FRIDAY(5), SATURDAY(6),
-		SUNDAY(7);
+	// Builder class migrated from the Test package
```
Remove this comment. The reason why it was migrated is not relevant for the code itself.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Schedule.java:
```diff
@@ -19,38 +19,100 @@
+			} else {
+				schedule.timeOnTypeInt = this.fixedTimeOn;
+				schedule.timeOffTypeInt = this.fixedTimeOff;
+			}
```
I don't understand what this if-statement does. Add a comment or change the code.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Schedule.java:
```diff
@@ -19,38 +19,100 @@
+			schedule.dayInt = this.dayInt;
+
+			if (this.fixedTimeOn < 0 || this.fixedTimeOff < 0) {
+				schedule.timeOnTypeInt = this.fixedTimeInt;
+				schedule.timeOffTypeInt = this.fixedTimeInt;
```
Why is it necessary to copy the same property to both timOnTypeInt and timeOffTypeInt?

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/Schedule.java:
```diff
@@ -19,38 +19,100 @@
+			return this;
+		}
+
+		public ScheduleBuilder isEnabled(boolean isEnabled) {
```
Rename the parameter and the property to "enabled".

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/ScheduleDay.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Map;
+
+public enum ScheduleDay {
+	// Meant to replace dayInt for schedules
```
Remove this comment. Such a comment could be part of a commit message.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/ScheduleDay.java:
```diff
@@ -0,0 +1,33 @@
+		return (ScheduleDay) map.get(dayInt);
+	}
+
+	public int getDayInt() {
```
Do you really need this method? Or can you convert an int to a ScheduleDay in the places, where you call getDayInt().

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculator.java:
```diff
@@ -20,54 +20,42 @@
-				logger.info("No relay found, number {}", relayNr);
 			}
+		} catch (Exception e) {
+			throw new SwitchingMomentCalculationException(
```
Add the original exception as second parameter to the SwitchingMomentCalculationException constructor.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/SwitchingMomentCalculator.java:
```diff
@@ -76,91 +64,84 @@ private SwitchingMoment calculateSwitchingMoment(int relayNr, Schedule schedule,
 		SwitchingMoment actualSwitchingMoment = new SwitchingMoment(relayNr, null, triggerAction);
 		int scheduledOccurence = schedule.getDayInt();
```
Change schedule.dayInt from integer to ScheduleDay (and change the name of the field accordingly). This makes the switch statement easier.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/TriggerType.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Map;
+
+public enum TriggerType {
+	// Meant to replace TimeType for schedules
```
Remove this comment.

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/TriggerType.java:
```diff
@@ -0,0 +1,32 @@
+		return (TriggerType) map.get(timeTypeInt);
+	}
+
+	public int getDayInt() {
```
The name of the method is wrong (copy-paste?) and do you really need it?

rlemmers commented on iec61850serversimulator/src/main/java/com/cgi/iec61850serversimulator/ScheduleDay.java:
```diff
@@ -0,0 +1,33 @@
+	SATURDAY(6), SUNDAY(7);
+
+	private int dayInt;
+	private static Map map = new HashMap<>();
```
Change this to Map<Integer, ScheduleDay>. It makes clear what kind of keys and values are stored in the map. It also enables simplification of the valueOf(..) method to "return map.get(dayInt);"

The same applies to the other enum.

kevin-wijnen committed:
Delete hs_err_pid20174.log

kevin-wijnen committed:
Pull Request Feedback implemented (mostly)

- Removed unnecessary log file
- Removed unnecessary comments
- Added clarification comments/recoding of the Schedule class'
If-statement
- Renamed enabled parameter in Schedule
- Removed unnecessary functions in the Enums
- Added original exception message to
SwitchingMomentCalculationException
- Added ScheduleDay to Schedule as property (but further
implementation/migration needs to get looked at, at a later date. dayInt
is still currently being used)

rlemmers committed:
78 ~ Minor cleanup
